src/main/scala/is/hail/expr/AST.scala

```diff
@@ -822,33 +815,41 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
     case (t, _, _) => FunctionRegistry.call(method, lhs +: args, t +: args.map(_.`type`))
   }
 
-  def toIR(agg: Option[String] = None): Option[IR] = (lhs.`type`, method, args: IndexedSeq[AST]) match {
-    case (_: TArray, "[]", IndexedSeq(rhs)) =>
-      for {
-        a <- lhs.toIR(agg)
-        i <- rhs.toIR(agg)
-      } yield ir.ArrayRef(a, i, `type`)
-    case (_: TAggregable, "sum", IndexedSeq()) =>
-      for {
-        a <- lhs.toIR(agg)
-        elementType = a.typ.asInstanceOf[TAggregable].elementType
-      } yield ir.ApplyAggOp(a, ir.Sum(), Seq(), elementType)
-    case (_: TAggregable, m, IndexedSeq(Lambda(_, name, body))) =>
-      for {
-        a <- lhs.toIR(agg)
-        b <- body.toIR(agg)
-        result <- optMatch(m) {
-          case "map" => ir.AggMap(a, name, b, `type`.asInstanceOf[TAggregable])
-          case "filter" => ir.AggFilter(a, name, b, `type`.asInstanceOf[TAggregable])
-          case "flatMap" => ir.AggFlatMap(a, name, b, `type`.asInstanceOf[TAggregable])
-        }
-      } yield result
-    case (_: TTuple, "[]", IndexedSeq(rhs)) =>
-      val idx = rhs.asInstanceOf[Const].value.asInstanceOf[Int]
-      for {
-        tup <- lhs.toIR(agg)
-      } yield ir.GetTupleElement(tup, idx, `type`)
-    case _ => None
+  def toIR(agg: Option[String] = None): Option[IR] = {
+    (lhs.`type`, method, args: IndexedSeq[AST]) match {
+      case (_: TArray, "[]", IndexedSeq(rhs)) =>
+        for {
+          a <- lhs.toIR(agg)
+          i <- rhs.toIR(agg)
+        } yield ir.ArrayRef(a, i, `type`)
+      case (_: TAggregable, "sum", IndexedSeq()) =>
+        for {
+          a <- lhs.toIR(agg)
+          elementType = a.typ.asInstanceOf[TAggregable].elementType
+        } yield ir.ApplyAggOp(a, ir.Sum(), Seq(), elementType)
+      case (_: TAggregable, m, IndexedSeq(Lambda(_, name, body))) =>
+        for {
+          a <- lhs.toIR(agg)
+          b <- body.toIR(agg)
+          result <- optMatch(m) {
+            case "map" => ir.AggMap(a, name, b, `type`.asInstanceOf[TAggregable])
+            case "filter" => ir.AggFilter(a, name, b, `type`.asInstanceOf[TAggregable])
+            case "flatMap" => ir.AggFlatMap(a, name, b, `type`.asInstanceOf[TAggregable])
+          }
+        } yield result
+      case (_: TTuple, "[]", IndexedSeq(rhs)) =>
+        val idx = rhs.asInstanceOf[Const].value.asInstanceOf[Int]
+        for {
+          tup <- lhs.toIR(agg)
+        } yield ir.GetTupleElement(tup, idx, `type`)
+      case (_, _, IndexedSeq(l: Lambda)) =>
+        IRFromAST.convertLambda(method, lhs, l, agg)
```

Comment:
You should just process the lambda here in place like we do above with the aggregables.  The IR has the array map operator and the lambda "fused" with a single ArrayMap operator, but you can match the two pieces together here and treat them like a monolithic unit.